### ISSUE TITLE
How to figure out the ideal number of contrastive steps

### ISSUE THREAD

**Issue by vahuja4** [huggingface/setfit#380]
I am seeing that utterances belonging to different classes in my dataset are lying close to each other in the embedding space. For example, when I try to find utterances similar to `I need to change beneficiary` **[updateBeneficiary]**, I get utterances from other classes too, like `I need to remove beneficiary`**[removeBeneficiary]** and `Want to add beneficiary to my account `**[addBeneficiary]**.  Also, all the beneficiary related datapoints are being predicted as **[addBeneficiary]** regardless of the groundtruth. Representation in the training set is as follows: `{"addBeneficiary":70, "updateBeneficiary":21, "removeBeneficiary":29}`. Here are my questions:
1. Is the setfit approach sensitive to this kind of class imbalance?
2. Can this be solved by increasing the number of contrastive examples, currently set to `15`? But increasing this would lead to overfitting?

**Comment by grofte** [comment]
This sounds really interesting. Did you make progress? Have you tried Optuna to tweak the hyperparameters? I'm thinking that it might be the logistic regression in the head that needs tuning rather than the contrastive iterations on the body.

**Comment by tomaarsen** [comment]
After #439 is merged & released this week, then you'll be able to track evaluation loss during training. This might help guide your choice for the (new) `max_steps` and `num_epochs` parameters. You will also be able to use an `EarlyStoppingCallback` from `transformers` with a `load_best_model_at_end=True` argument to load the model which has the best evaluation performance rather than the model with the best training performance. 
This should help avoid overfitting.
